config/initializers/diaspora_federation.rb

```diff
@@ -63,64 +63,95 @@
       person_entity.save!
     end
 
-    on :fetch_private_key_by_diaspora_id do |diaspora_id|
+    on :fetch_private_key do |diaspora_id|
       key = Person.where(diaspora_handle: diaspora_id).joins(:owner).pluck(:serialized_private_key).first
-      OpenSSL::PKey::RSA.new key unless key.nil?
+      OpenSSL::PKey::RSA.new(key) unless key.nil?
     end
 
-    on :fetch_author_private_key_by_entity_guid do |entity_type, guid|
-      key = entity_type.constantize.where(guid: guid).joins(author: :owner).pluck(:serialized_private_key).first
-      OpenSSL::PKey::RSA.new key unless key.nil?
+    on :fetch_public_key do |diaspora_id|
+      key = Person.find_or_fetch_by_identifier(diaspora_id).serialized_public_key
+      OpenSSL::PKey::RSA.new(key) unless key.nil?
     end
 
-    on :fetch_public_key_by_diaspora_id do |diaspora_id|
-      key = Person.where(diaspora_handle: diaspora_id).pluck(:serialized_public_key).first
-      OpenSSL::PKey::RSA.new key unless key.nil?
+    on :fetch_related_entity do |entity_type, guid|
+      entity = entity_type.constantize.find_by(guid: guid)
+      Diaspora::Federation::Entities.related_entity(entity) if entity
     end
 
-    on :fetch_author_public_key_by_entity_guid do |entity_type, guid|
-      key = entity_type.constantize.where(guid: guid).joins(:author).pluck(:serialized_public_key).first
-      OpenSSL::PKey::RSA.new key unless key.nil?
+    on :queue_public_receive do |xml, legacy=false|
+      Workers::ReceivePublic.perform_async(xml, legacy)
     end
 
-    on :entity_author_is_local? do |entity_type, guid|
-      entity_type.constantize.where(guid: guid).joins(author: :owner).exists?
-    end
-
-    on :fetch_entity_author_id_by_guid do |entity_type, guid|
-      entity_type.constantize.where(guid: guid).joins(:author).pluck(:diaspora_handle).first
-    end
-
-    on :queue_public_receive do |xml|
-      Workers::ReceiveUnencryptedSalmon.perform_async(xml)
-    end
-
-    on :queue_private_receive do |guid, xml|
+    on :queue_private_receive do |guid, xml, legacy=false|
       person = Person.find_by_guid(guid)
 
-      if person.nil? || person.owner_id.nil?
-        false
-      else
-        Workers::ReceiveEncryptedSalmon.perform_async(person.owner.id, xml)
-        true
+      (person.present? && person.owner_id.present?).tap do |user_found|
+        Workers::ReceivePrivate.perform_async(person.owner.id, xml, legacy) if user_found
       end
     end
 
-    on :receive_entity do
-      # TODO
+    on :receive_entity do |entity, recipient_id|
+      case entity
+      when DiasporaFederation::Entities::AccountDeletion
+        Diaspora::Federation::Receive.account_deletion(entity)
+      when DiasporaFederation::Entities::Retraction
+        Diaspora::Federation::Receive.retraction(entity, recipient_id)
+      else
+        persisted = case entity
```

Comment:
A hash, moving it out of sight of the glue code and `public_send` is probably acceptable here.

```rb
Diaspora::Federation::Receive.perform(entity)

def self.perform(entity)
  receiver = ENTITY_RECEIVERS.fetch(entity.class) { raise ... }
  public_send(receiver, entity)
end